### PR TITLE
docs: Search and Replace docs

### PR DESCRIPTION
**Pull request by phanimarupaka**
Docs for Search and Replace for next branch. Follow up to https://github.com/GoogleContainerTools/kpt/pull/1231. The docs were verified during Fishfood and approved but closed the PR as it should not be merged into master back then.

### CODE REVIEW

File: internal/docs/generated/cfgdocs/docs.go [hunk 1]
```diff
@@ -1,17 +1,3 @@
-// Copyright 2019 Google LLC
```
Comment (mortent) [inline internal/docs/generated/cfgdocs/docs.go:1]: We should add this back in.

File: site/content/en/reference/cfg/search/_index.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  Search and optionally replace fields across all resources.
+---
+
+<!--mdtogo:Short
+    Search and optionally replace fields across all resources.
+-->
+
+{{% pageinfo color="primary" %}}
+
+This feature is under development. Please set environment variable
```
Comment (mortent) [inline site/content/en/reference/cfg/search/_index.md:16]: In this branch we should make this functionality available without requiring the env variable.

File: internal/docs/generated/cfgdocs/docs.go (hunk 1 above)
Comment (phanimarupaka) [reply internal/docs/generated/cfgdocs/docs.go]: @mortent Done.

File: site/content/en/reference/cfg/search/_index.md (hunk 2 above)
Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Done.

**Review by mortent — APPROVED** [review]

File: internal/cmdsearch/cmdsearch.go [hunk 3]
```diff
@@ -57,11 +57,6 @@ func NewSearchRunner(name string) *SearchRunner {
 	return r
 }
 
-const shortMessage = `Search and optionally replace fields across all resources.
```
Comment (frankfarzan) [inline internal/cmdsearch/cmdsearch.go:60]: Why remove this?

File: site/content/en/reference/cfg/search/_index.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+<!--mdtogo:Examples-->
+
+```sh
+# Matches fields with value "3":
+$ kpt cfg search DIR/ --by-value 3
+```
+
+```sh
+# Matches fields with value prefixed by "nginx-":
+$ kpt cfg search DIR/ --by-value-regex ngnix-.*
```
Comment (frankfarzan) [inline site/content/en/reference/cfg/search/_index.md:78]: I had comments int the doc about quoting the values so it's not consumed by shell

File: site/content/en/reference/cfg/search/_index.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+```
+
+```sh
+# Set namespaces for all resources to "bookstore", even namespace is not set on a resource:
+$ kpt cfg search DIR/ --by-path metadata.namespace --put-literal bookstore
+```
+
+```sh
+# Put the setter pattern as a line comment for matching fields.
+$ kpt cfg search DIR/ --by-value my-project-id-foo --put-pattern '${project-id}-foo'
```
Comment (frankfarzan) [inline site/content/en/reference/cfg/search/_index.md:98]: how about referencing capture groups when using by regex

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: internal/cmdsearch/cmdsearch.go (hunk 3 above)
Comment (phanimarupaka) [reply internal/cmdsearch/cmdsearch.go:60]: They are auto-generated from mdtogo and added here. https://github.com/GoogleContainerTools/kpt/pull/1456/files#diff-d16bb5b078337f35bfbb11af3240a9477e2d7b5972739f56cf9b137beb90d7d9R109

File: site/content/en/reference/cfg/search/_index.md (hunk 4 above)
Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Thanks. Added single quotes.

File: site/content/en/reference/cfg/search/_index.md (hunk 5 above)
Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: This is the simple case of adding pattern. Added example for regex case. Thanks

Comment (frankfarzan) [reply site/content/en/reference/cfg/search/_index.md]: Can you elaborate? 

Will something like this be supported?

$ kpt cfg search DIR/ --by-value-regex 'project-id-(\w)-(\w)' --put-pattern 'my-${project-id}-${1}-${2}'

Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Yes. This works but the command would be 

$ kpt cfg search DIR/ --by-value-regex 'my-lazy-koala-(\w)-(\w)' --put-pattern 'my-${project-id}-(\w)-(\w)'

This means match all the field values with pattern 'my-lazy-koala-(\w)-(\w)' add the setter pattern comment deriving the setter value.

For fields with value 
my-lazy-koala-foo-bar
...
my-lazy-koala-baz-daz

Setter pattern will be added
my-lazy-koala-foo-bar # kpt-set: my-{project-id}-foo-bar
...
my-lazy-koala-baz-daz # kpt-set: my-{project-id}-baz-daz

Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Updated the docs with recent changes.

File: site/content/en/reference/cfg/search/_index.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+value of the field by default without requiring start (^) and end ($) characters.
+
+--by-path
+Match by path expression of a field. Path expressions are used to deeply navigate
+and match particular yaml nodes. Please note that the path expressions are not
+regular expressions.
+
+--put-value
+Set or update the value of the matching fields. Input can be a pattern for which
+the capture groups are resolved using --by-value-regex input.
```
Comment (frankfarzan) [inline site/content/en/reference/cfg/search/_index.md:59]: Should mention that only numbered capture groups are resolved. (There are also named groups which may be conflated with setters)

File: site/content/en/reference/cfg/search/_index.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+--put-comment
+Set or update the line comment for matching fields. Input can be a pattern for
+which the capture groups are resolved using --by-value-regex input.
+
+--recurse-subpackages
```
Comment (frankfarzan) [inline site/content/en/reference/cfg/search/_index.md:65]: this is the default behavior right?

**Review by frankfarzan — APPROVED** [review]
minor questions

File: site/content/en/reference/cfg/search/_index.md (hunk 7 above)
Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Yes. Updated the default.

File: site/content/en/reference/cfg/search/_index.md (hunk 6 above)
Comment (phanimarupaka) [reply site/content/en/reference/cfg/search/_index.md]: Done.

**Pull request merged**